x/evm/keeper/keeper.go

```diff
@@ -201,40 +207,68 @@ func (k *Keeper) CalculateNextNonce(ctx sdk.Context, addr common.Address, includ
 	}
 
 	// get the pending nonces (nil is fine)
-	pending := k.pendingNonces[addr.Hex()]
+	pending := k.pendingTxs[addr.Hex()]
 
 	// Check each nonce starting from latest until we find a gap
 	// That gap is the next nonce we should use.
 	for ; ; nextNonce++ {
 		// if it's not in pending, then it's the next nonce
-		if _, found := sort.Find(len(pending), func(i int) int { return uint64Cmp(nextNonce, pending[i]) }); !found {
+		if _, found := sort.Find(len(pending), func(i int) int { return uint64Cmp(nextNonce, pending[i].Nonce) }); !found {
 			return nextNonce
 		}
 	}
 }
 
 // AddPendingNonce adds a pending nonce to the keeper
-func (k *Keeper) AddPendingNonce(key tmtypes.TxKey, addr common.Address, nonce uint64) {
+func (k *Keeper) AddPendingNonce(key tmtypes.TxKey, addr common.Address, nonce uint64, priority int64) {
 	k.nonceMx.Lock()
 	defer k.nonceMx.Unlock()
 
 	addrStr := addr.Hex()
 	if existing, ok := k.keyToNonce[key]; ok {
-		if existing.nonce != nonce {
-			fmt.Printf("Seeing transactions with the same hash %X but different nonces (%d vs. %d), which should be impossible\n", key, nonce, existing.nonce)
+		if existing.Nonce != nonce {
+			fmt.Printf("Seeing transactions with the same hash %X but different nonces (%d vs. %d), which should be impossible\n", key, nonce, existing.Nonce)
 		}
-		if existing.address != addr {
-			fmt.Printf("Seeing transactions with the same hash %X but different addresses (%s vs. %s), which should be impossible\n", key, addr.Hex(), existing.address.Hex())
+		if existing.Address != addr {
+			fmt.Printf("Seeing transactions with the same hash %X but different addresses (%s vs. %s), which should be impossible\n", key, addr.Hex(), existing.Address.Hex())
 		}
 		// we want to no-op whether it's a genuine duplicate or not
 		return
 	}
-	k.keyToNonce[key] = &addressNoncePair{
-		address: addr,
-		nonce:   nonce,
+	for _, pendingTx := range k.pendingTxs[addrStr] {
+		if pendingTx.Nonce == nonce {
+			if priority > pendingTx.Priority {
+				// replace existing tx
+				delete(k.keyToNonce, pendingTx.Key)
```

Comment:
do we also need to update tendermint's pending tx list somewhere here? i don't think so but wanted to double check